### PR TITLE
cMAB Fast Update via Variational Inference

### PR DESCRIPTION
**Pull request by shaharbar1**
### Changes
 * Edited BaseBayesianLogisticRegression and inheritors on model.py to support variational inference by adding fast_inference control parameter on class attributes and adding control arguments on update method.
 * Edited update UTs on test_cmab.py to support new inference mode.
 * Edited cMABs cold start function tto support new inference mode.
 * Removed redundant test_execution_time.py.
 * Edited version on pyproject.toml.

### CODE REVIEW

File: pybandits/cmab.py [hunk 1]
```diff
@@ -355,7 +358,7 @@ def create_cmab_bernoulli_cold_start(
     """
     actions = {}
     for a in set(action_ids):
```
Comment (adarmiento) [inline pybandits/cmab.py:360]:
```suggestion
    for action_id in set(action_ids):
```

File: pybandits/cmab.py [hunk 2]
```diff
@@ -347,6 +348,8 @@ def create_cmab_bernoulli_cold_start(
     default_action: Optional[ActionId]
         The default action to select with a probability of epsilon when using the epsilon-greedy approach.
         If `default_action` is None, a random action from the action set will be selected with a probability of epsilon.
+    fast_inference: bool, defaults to False
+        Whether to utilize MCMC (False) or variational inference (True) for the Bayesian inference on update
```
Comment (adarmiento) [inline pybandits/cmab.py:352]:
```suggestion
        Whether to utilize standard MCMC (False) or fast variational inference (True) for the Bayesian inference on update
```

File: pybandits/model.py [hunk 3]
```diff
@@ -313,14 +314,15 @@ def update(
...
-        chains=2,
-        init="adapt_diag",
-        cores=2,
-        target_accept=0.95,
-        progressbar=False,
-        return_inferencedata=False,
+        tune: Optional[int] = 500,
+        draws: int = 1000,
+        chains: Optional[int] = 2,
+        init: Optional[str] = "adapt_diag",
```
Comment (adarmiento) [inline pybandits/model.py:320]: This could be an enum somewhere

Comment (shaharbar1) [reply pybandits/model.py]: Thanks for your comments mate :-)

File: pybandits/cmab.py [hunk 4]
```diff
@@ -347,15 +358,21 @@ def create_cmab_bernoulli_cold_start(
     default_action: Optional[ActionId]
         The default action to select with a probability of epsilon when using the epsilon-greedy approach.
         If `default_action` is None, a random action from the action set will be selected with a probability of epsilon.
+    update_method: UpdateMethods, defaults to MCMC
+        Whether to utilize MCMC or variational inference for the Bayesian inference on update.
```
Comment (dariodandrea) [inline pybandits/cmab.py:362]: Let's also indicate where to find the list of available updated_method (in mode
```
update_method: UpdateMethods, defaults "MCMC"
   The strategy for computing posterior quantities of the Bayesian models in the update function. Such as Markov 
   chain Monte Carlo ("MCMC") or Variational Inference ("VI"). Check UpdateMethods in pybandits.model for the 
   full list. 
```

File: pybandits/model.py [hunk 5]
```diff
@@ -231,16 +234,61 @@ class BaseBayesianLogisticRegression(Model):
...
+        progressbar=False,
+        return_inferencedata=False,
+    )
+    _default_variational_inference_kwargs = dict(method="advi")
+
+    @model_validator(mode="after")
+    def arrange_update_kwargs(self):
+        if self.update_kwargs is None:
+            self.update_kwargs = self._default_update_kwargs
+        if self.update_method:
```
Comment (dariodandrea) [inline pybandits/model.py:268]: Update according to the new update_method 'MCMC' or 'VI'

File: pybandits/model.py [hunk 6]
```diff
@@ -231,16 +234,61 @@ class BaseBayesianLogisticRegression(Model):
...
+        else:
+            self.update_kwargs = {**self._default_mcmc_kwargs, **self.update_kwargs}
+        return self
+
+    @classmethod
+    def _stable_sigmoid(cls, x: Union[np.ndarray, TensorVariable]) -> Union[np.ndarray, TensorVariable]:
+        """
+        Vectorized sigmoid function that avoids overflow and underflow.
+        Compatible with both numpy and PyMC3 tensors.
+        Parameters
```
Comment (dariodandrea) [inline pybandits/model.py:279]: please keep an empty line here before Parameters

File: pybandits/model.py [hunk 7]
```diff
@@ -231,16 +234,62 @@ class BaseBayesianLogisticRegression(Model):
...
-    alpha: StudentT
+    alpha : StudentT
         Student's t-distribution of the alpha coefficient.
-    betas: StudentT
+    betas : StudentT
         Student's t-distributions of the betas coefficients.
-    params_sample: Dict
-        Parameters for the function pymc.sample()
+    update_method : UpdateMethods, defaults to "MCMC"
+        Whether to utilize standard MCMC (False) or faster variational inference (True).
```
Comment (dariodandrea) [inline pybandits/model.py:242]: docs here is outdated

File: pybandits/model.py [hunk 8]
```diff
@@ -392,12 +427,14 @@ class BayesianLogisticRegression(BaseBayesianLogisticRegression):
...
```
Comment (dariodandrea) [inline pybandits/model.py:435]: docs here is outdated

File: pybandits/model.py [hunk 9]
```diff
@@ -417,16 +454,20 @@ class BayesianLogisticRegressionCC(BaseBayesianLogisticRegression):
         Student's t-distribution of the alpha coefficient.
     betas: StudentT
         Student's t-distributions of the betas coefficients.
-    params_sample: Dict
-        Parameters for the function pymc.sample()
+    update_method : UpdateMethods, defaults to "MCMC"
+        Whether to utilize standard MCMC (False) or faster variational inference (True).
```
Comment (dariodandrea) [inline pybandits/model.py:458]: docs here is outdated

File: pybandits/model.py [hunk 10]
```diff
@@ -441,17 +482,29 @@ def create_bayesian_logistic_regression_cold_start(n_betas: PositiveInt) -> Baye
     n_betas : PositiveInt
         The number of betas of the Bayesian Logistic Regression model. This is also the number of features expected
         after in the context matrix.
+    update_method : UpdateMethods, defaults to "MCMC"
+        Whether to utilize standard MCMC (False) or faster variational inference (True).
```
Comment (dariodandrea) [inline pybandits/model.py:486]: docs here is outdated

File: pybandits/model.py [hunk 11]
```diff
@@ -344,40 +380,39 @@ def update(
...
-                **kwargs,
-            )
+            if self.update_method == "VI":
+                # variational inference
+                update_kwargs = self.update_kwargs.copy()
+                approx = fit(method=update_kwargs.pop("method"))
+                trace = approx.sample(**update_kwargs)
+            else:
+                # MCMC
+                trace = sample(**self.update_kwargs)
```
[the remark this answers is not in the document]
Comment (dariodandrea) [reply pybandits/model.py:406]: even if we never allow to init a cmab with unsupported updated method, right?

File: pybandits/model.py [hunk 12]
```diff
@@ -231,16 +234,62 @@ class BaseBayesianLogisticRegression(Model):
...
+    )
+    _default_variational_inference_kwargs = dict(method="advi")
+
+    @model_validator(mode="after")
+    def arrange_update_kwargs(self):
+        if self.update_kwargs is None:
+            self.update_kwargs = self._default_update_kwargs
+        if self.update_method == "VI":
+            self.update_kwargs = {**self._default_variational_inference_kwargs, **self.update_kwargs}
+        else:
```
Comment (dariodandrea) [inline pybandits/model.py:270]: Sme comment as in line 399 if-else statement

File: pybandits/model.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (shaharbar1) [reply pybandits/model.py:406]: Agree. It's good to catch future interface updates errors.

**Review by dariodandrea — APPROVED** [review]
Great job @shaharbar1 !

**Pull request merged**